pkg/services/sqlstore/annotation.go

```diff
@@ -268,3 +268,53 @@ func (r *SQLAnnotationRepo) Delete(params *annotations.DeleteParams) error {
 		return nil
 	})
 }
+
+func (r *SQLAnnotationRepo) FindTags(query *annotations.TagsQuery) ([]*annotations.TagsDTO, error) {
+	if query.Limit == 0 {
+		query.Limit = 100
+	}
+
+	var sql bytes.Buffer
+	params := make([]interface{}, 0)
+	tagKey := `tag.` + dialect.Quote("key")
+	tagValue := `tag.` + dialect.Quote("value")
+
+	sql.WriteString(`
+		SELECT
+			` + tagKey + `,
+			` + tagValue + `,
+			count(*) as count
+		FROM tag
+		INNER JOIN annotation_tag ON tag.id = annotation_tag.tag_id
+		INNER JOIN annotation ON annotation_tag.annotation_id = annotation.id
```

Comment:
This will only be used when editing annotations for a dashboard and when adding/editing annotations in the UI right? Should be fine performance wise, but the inner joins operate on more data. Not sure if using JOIN vs EXISTS would improve things, like `where exists(select 1 from annotation where id = annotation_tag.annotation_id`. For the count join on annotation_tag should/could be enough?

Reply:
It will also be used when creating/editing annotation queries for the built-in Grafana data source.
![image](https://user-images.githubusercontent.com/562238/123900837-1959ac00-d96a-11eb-9821-efc555a90fbb.png)

Reply:
Excellent suggestion, I've made changes.

Reply:
I dig a bit deeper into this for SQLLite only:

For this SQL:
```sql
select
    tag.key,
    tag.value,
    count(*) as count
from tag
inner join annotation_tag on tag.id = annotation_tag.tag_id
inner join annotation on annotation_tag.annotation_id = annotation.id
where annotation.org_id = 1 and tag.key like '%%' or tag.value like '%%'
group by tag.key, tag.value
order by tag.key, tag.value
```

The explain plan shows a mixed plan with one full scan and 2 index scans:
![image](https://user-images.githubusercontent.com/562238/123908237-6e4fef00-d977-11eb-9a53-ebb41e56ee23.png)


For this SQL:
```sql
select
    tag.key,
    tag.value,
    count(*) as count
from tag
         inner join annotation_tag on tag.id = annotation_tag.tag_id
where exists(select 1 from annotation where annotation.id = annotation_tag.annotation_id and annotation.org_id = 1)
and tag.key like '%%' or tag.value like '%%'
group by tag.key, tag.value
order by tag.key, tag.value
```

The explain plan shows a mixed plan with one full scan and 2 index scans and a subquery:
![image](https://user-images.githubusercontent.com/562238/123908589-02ba5180-d978-11eb-9ab1-2cd90790fe1b.png)

I'm not an expert on this at all but I can see how the later would be better with a lot of annotations. 

But isn't the biggest issue in both cases the Full Scan which I think is caused by the `tag.key like '%%' or tag.value like '%%'`?

Reply:
I changed to the later SQL statement.

Reply:
Nice. I think you're missing parenthesis around the or, like `and (tag.key like '%%' or tag.value like '%%')`. Databases cannot optimize for like-queries beginning with `%` - it always need to do a full table scan. I think in the code looked like you did `like '<search term>%`? When the search term is empty, might not want to apply the filter at all? I would like to check out the plan for mysql so gonna try that out since MySQL or Postgres are normally what are being used in production setups and cloud.

Reply:
Good catch with the parenthesis, will change ASAP.

Reply:
Nevermind, in the code I have the parenthesis, but when I executed the plans I didn't. 🙈

Reply:
New results for SQL Lite, now with the parenthesis. Seems like the first query doesn't need a full scan any longer?

```sql
select
    tag.key,
    tag.value,
    count(*) as count
from tag
inner join annotation_tag on tag.id = annotation_tag.tag_id
inner join annotation on annotation_tag.annotation_id = annotation.id
where annotation.org_id = 1
and (tag.key like '%%' or tag.value like '%%')
group by tag.key, tag.value
order by tag.key, tag.value
```
![image](https://user-images.githubusercontent.com/562238/123951342-f993a980-d9a4-11eb-9914-1867f4c9211f.png)

```sql
select
    tag.key,
    tag.value,
    count(*) as count
from tag
         inner join annotation_tag on tag.id = annotation_tag.tag_id
where exists(select 1 from annotation where annotation.id = annotation_tag.annotation_id and annotation.org_id = 1)
and (tag.key like '%%' or tag.value like '%%')
group by tag.key, tag.value
order by tag.key, tag.value
```
![image](https://user-images.githubusercontent.com/562238/123951463-1a5bff00-d9a5-11eb-95f7-65d664fbde0c.png)

Reply:
Do you think this is a stopper for this PR @marefr or can I optimize this in a later PR?

Reply:
Definitely no stopper